File: java/FractiontoRecurringDecimal.java
/*
 * Filename: FractiontoRecurringDecimal.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Wed 15 Jul 2015 03:17:06 PM EDT
 */

public class Solution {
	public String fractionToDecimal(int numerator, int denominator) {
		if (numerator == 0) {
			return "0";
		}

		if (denominator == 0) {
			return "";
		}
		
		String res = "";
		if ((numerator > 0 && denominator < 0)
				|| (numerator < 0 && denominator > 0)) {
			res += "-";
		}

		long num = Math.abs((long) numerator);
		long denom = Math.abs((long) denominator);

		System.out.println(num);
		System.out.println(denom);

		long quotient = num / denom;
		num = (num % denom) * 10;
		res += quotient;

		if (num == 0) {
			return res;
		}

		HashMap<Long, Integer> map = new HashMap<Long, Integer>();
		res += ".";

		while (num != 0) {
			if (map.containsKey(num)) {
				int i = map.get(num);
				return res.substring(0, i) + "(" + res.substring(i) + ")";
			}
			
			map.put(num, res.length());
			quotient = num / denom;
			num = (num % denom) * 10;
			res += quotient;
		}

		return res;
	}
}

File: cpp/fraction_to_recurring_decimal.cpp
/*
 * Filename: fraction_to_recurring_decimal.cpp
 * Author: Jun Xu
 * Email: junx@cs.bu.edu
 * Created Time: Mon 16 Nov 2015 10:05:16 PM EST
 * Description: 
 */

#include <iostream>
#include <map>

using namespace std;

class Solution {
	public:
		string fractionToDecimal(int numerator, int denominator) {
			if (numerator == 0 || denominator == 0)
			{
				return "0";
			}
			
			string res = "";
			if ((numerator < 0) ^ (denominator < 0))
			{
				res += "-";
			}

			long long int num_long = numerator;
			long long int denom_long = denominator;
			num_long = abs(num_long);
			denom_long = abs(denom_long);

			res += to_string(num_long / denom_long);
			int residual = num_long % denom_long;

			if (!residual)
			{
				return res;
			}

			res += ".";
			map<int, int> map;

			while (residual)
			{
				if (map.find(residual) != map.end())
				{
					int i = map[residual];
					//res.insert(i, 1, '(');
					//return res + ")";
					return res.substr(0, i) + "(" + res.substr(i) + ")";
				}

				map[residual] = res.length();
				res += to_string(residual * 10 / denom_long);
				residual = residual * 10 % denom_long;
			}

			return res;
		}
};
